### PR TITLE
Added initial API style guidelines

### PR DESCRIPTION
**Pull request by oogg**
@archgrove @dcoxall @samhandev @steve-codes

### CODE REVIEW

**Comment by archgrove** [comment]
Something about the use of appropriate content types (at the very least, `application/json`) and character encodings?

File: api.md [hunk 1]
```diff
@@ -1,13 +1,102 @@
...
+
+The `GET` verb should not effect the system, and should not change response on subsequent requests (unless the underlying data changes), i.e. it should be idempotent. Exceptions to 'changing the system' are typically instrumentation/logging-related.
+
+Use `POST` to create one or multiple Resources. Use `PUT` to update or create a single Resource. Use `DELETE` to delete a Resource.
+
+
+### Actions or Complex operations
+Avoid actions or verbs in the URL. In special cases where these are needed, such as when a resource is moving to another resource (so state is changing against two Resources), perform the operation using a `POST`.
+e.g. 
+    `POST` `/basket/items/{id}/saveforlater`
```
Comment (steve-codes) [inline api.md]: Do we need consistency in resource naming?
Here you using the concatenated `saveforlater`, where in earlier examples you're using dash separated `my-offers`

File: api.md [hunk 2]
```diff
@@ -1,13 +1,102 @@
...
 
   > The request has been accepted for processing, but the processing has not
   > been completed.
 
-- Use `422` for validation errors, as described in [RFC 2918](http://tools.ietf.org/html/rfc4918#section-11.2):
+#### Authentication and Authorization error codes
+- `401 Unauthorized`: Request failed because user is not authenticated
+- `403 Forbidden`: Request failed because user does not have authorization to access a specific resource
+
+#### Make errors useful
```
Comment (steve-codes) [inline api.md]: Should we be returning a standardised model for the body of an error response? Is this something we want to define in this document?

File: api.md [hunk 3]
```diff
@@ -16,3 +105,55 @@
...
+
+      {
+        "totalQuantity": 18.99,
+        "customerId": "123456...",
+        // ...
+      }
+
+This approach makes it possible to inline more information about the related resource without having to change the structure of the response or introduce more top-level response fields.
+  
+### Returning errors
```
Comment (steve-codes) [inline api.md]: Can we move this section to be closer to the HTTP Status codes section?

File: api.md [hunk 4]
```diff
@@ -16,3 +105,55 @@
...
+
+
+## Todo
+- Detail the use of `Patch`
+- Health endpoints
+- Hypermedia pattern
+- Casing in attributes/fields (and use of abbreviation in naming)
+- Headers indicating latency from dependencies
+- `RateLimit-Remaining` header
+- Events / event sourcing
```
Comment (steve-codes) [inline api.md]: We should add "Authentication/Authrorisation" to the list of TODO. 
I think it's also worth thinking about what L10N and I18N looks like in reference to our API. Should things like amounts, prices etc by localised by the client or by the API?

File: api.md (hunk 3 above)
Comment (oogg) [inline api.md]: @steve-codes "Returning errors" closer to "http status codes"?

**Comment by oogg** [comment]
@steve-codes @archgrove @dcoxall @jabley 
Corrected consistency, added TODOs based on feedback (keeping more subjective topics for later commits)

File: api.md (hunk 3 above)
Comment (Neha-Nanda) [inline api.md]: There is one more status code - 204 No Content	GET, PUT, DELETE	 the request was processed successfully, but no response body is needed.
Are these status codes are  less used /commonly used status codes ?

File: api.md [hunk 5]
```diff
@@ -1,13 +1,102 @@
...
+
+#### Resources
+It is important that you have agreement from the API Product Owner and API community before reserving a Resource. Resources should reflect the customer's perspective, not necessarily the company's. 
+
+Collection resource names should be plural nouns, e.g. `/customers`. This helps visually disambiguate collections from singletons. e.g.:
+
+    /stores/v2
+
+### Leaky abstractions
+Avoid leaky abstractions in your URIs, field names and data structures. Ensure your clients don't care and don't know that your backend is for example a Mongo database, SAP, or an enterprise IBM product. That backend may change and you don't want to impact your (potentially many hundreds of) clients.
```
Comment (jabley) [inline api.md]: Maybe a note on the distinction between published IDs (that we commit to supporting and not breaking) and private IDs (that are a current implementation detail)?

**Comment by oogg** [comment]
To @Neha-Nanda's point above, there is some potential ambiguity around am I listing only the HTTP status codes that should be used, so what about the others? Will clarify (I am deliberately avoiding listing out every possible HTTP status code).

File: api.md [hunk 6]
```diff
@@ -1,13 +1,102 @@
...
+      -d '{
+        "skuId":1234,
+        "quantity":1
+        }' 
+
+### Downcase paths
+Use downcased and dash-separated path names e.g:
+      <server>/customers/my-offers
+
+### HTTP Verbs
```
Comment (jabley) [inline api.md]: [Use HTTP methods as Tim intended](https://www.gov.uk/service-manual/making-software/apis#use-http-methods-as-tim-intended) 😊

File: api.md (hunk 3 above)
Comment (jabley) [inline api.md]: I'm not sure we need to reproduce the HTTP spec here. But I'm a big fan of using `204 No Content` in general.

**Comment by jabley** [comment]
I like http://jsonapi.org/ for providing a consistent structure for JSON responses.

File: api.md [hunk 7]
```diff
@@ -1,13 +1,102 @@
...
-- Use `202` for times when the data is valid, but we've queued the persistence for background processing, as described in [RFC 7231](http://tools.ietf.org/html/rfc7231#section-6.3.3):
+
+## Foundations
+
+### TLS
+Always require secure connections with TLS without exception. 
+
+### URI components
+
+#### Version
```
Comment (jabley) [inline api.md]: *twitches*

Probably the pragmatic approach, but I dislike breaking things.

File: api.md [hunk 8]
```diff
@@ -1,13 +1,102 @@
...
+Collection resource names should be plural nouns, e.g. `/customers`. This helps visually disambiguate collections from singletons. e.g.:
+
+    /stores/v2
+
+### Leaky abstractions
+Avoid leaky abstractions in your URIs, field names and data structures. Ensure your clients don't care and don't know that your backend is for example a Mongo database, SAP, or an enterprise IBM product. That backend may change and you don't want to impact your (potentially many hundreds of) clients. 
+
+### Todo
+- Request IDs
+- Paging
```
Comment (jabley) [inline api.md]: Based off [RFC5005](https://tools.ietf.org/html/rfc5005), but with [consideration for cache behaviour and generating static responses](http://www.dehora.net/journal/2008/07/20/efficient-api-paging-count-down-not-up/).

**Comment by oogg** [comment]
Paired with @dcoxall
- added clarification around HTTP status codes (not being exhaustive).
- added `204`
- Ref to gov.uk link

**Comment by oogg** [comment]
Thanks all for feedback. Merging and looking forward to iterating

**Pull request merged**